{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6-final"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.6 64-bit",
   "metadata": {
    "interpreter": {
     "hash": "1ee38ef4a5a9feb55287fd749643f13d043cb0a7addaab2a9c224cbe137c0062"
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import math\n",
    "import random\n",
    "\n",
    "class Tetromino():\n",
    "\n",
    "    TYPES = [' ', 'I', 'O', 'T', 'S', 'Z', 'J', 'L']\n",
    "    TYPES_D = {\n",
    "        ' ': 0,\n",
    "        'I': 1,\n",
    "        'O': 2,\n",
    "        'T': 3,\n",
    "        'S': 4,\n",
    "        'Z': 5,\n",
    "        'J': 6,\n",
    "        'L': 7\n",
    "    }\n",
    "\n",
    "    def __init__(self, state):\n",
    "        self.state = np.array(state, dtype=np.uint8, copy=True)\n",
    "\n",
    "    @staticmethod\n",
    "    def ITetromino():\n",
    "\t    return Tetromino(\n",
    "            [\n",
    "                [1, 1, 1, 1]\n",
    "            ]\n",
    "        )\n",
    "\n",
    "    @staticmethod\n",
    "    def OTetromino():\n",
    "        return Tetromino(\n",
    "            [\n",
    "                [2, 2],\n",
    "                [2, 2]\n",
    "            ]\n",
    "        )\n",
    "\n",
    "    @staticmethod\n",
    "    def TTetromino():\n",
    "        return Tetromino(\n",
    "            [\n",
    "                [3, 3, 3],\n",
    "                [0, 3, 0]\n",
    "            ]\n",
    "        )\n",
    "\n",
    "    @staticmethod\n",
    "    def STetromino():\n",
    "        return Tetromino(\n",
    "            [\n",
    "                [0, 4, 4],\n",
    "                [4, 4, 0]\n",
    "            ]\n",
    "        )\n",
    "\n",
    "    @staticmethod\n",
    "    def ZTetromino():\n",
    "        return Tetromino(\n",
    "            [\n",
    "                [5, 5, 0],\n",
    "                [0, 5, 5]\n",
    "            ]\n",
    "        )\n",
    "\n",
    "    @staticmethod\n",
    "    def JTetromino():\n",
    "        return Tetromino(\n",
    "            [\n",
    "                [6, 6, 6],\n",
    "                [0, 0, 6]\n",
    "            ]\n",
    "        )\n",
    "\n",
    "    @staticmethod\n",
    "    def LTetromino():\n",
    "        return Tetromino(\n",
    "            [\n",
    "                [7, 7, 7],\n",
    "                [7, 0, 0]\n",
    "            ]\n",
    "        )\n",
    "\n",
    "    @staticmethod\n",
    "    def create(letter):\n",
    "        if letter.upper() in Tetromino.TYPES[1:]:\n",
    "            raise ValueError('No Tetromino of type {}'.format(letter))\n",
    "        return getattr(Tetromino, '{}Tetromino'.format(letter.upper()))()\n",
    "\n",
    "    def __str__(self):\n",
    "        return str(np.vectorize(Tetromino.TYPES.__getitem__)(self.state))\n",
    "\n",
    "    def __getitem__(self, key):\n",
    "        return self.state[key]\n",
    "\n",
    "    def copy(self):\n",
    "        return Tetromino(self.state)\n",
    "\n",
    "    def flat(self):\n",
    "        return self.state.flat\n",
    "\n",
    "    def width(self):\n",
    "        return self.state.shape[1]\n",
    "\n",
    "    def height(self):\n",
    "        return self.state.shape[0]\n",
    "\n",
    "    def rotate(self, change):\n",
    "        while change < 0:\n",
    "            change += 4\n",
    "        change = (change % 4)\n",
    "        if change == 0:\n",
    "            return None\n",
    "        elif change == 1:\n",
    "            self.rotate_right()\n",
    "        elif change == 2:\n",
    "            self.flip()\n",
    "        elif change == 3:\n",
    "            self.rotate_left()\n",
    "\n",
    "    def rotate_right(self):\n",
    "        self.state = np.rot90(self.state, 3)\n",
    "        return self\n",
    "\n",
    "    def rotate_left(self):\n",
    "        self.state = np.rot90(self.state, 1)\n",
    "        return self\n",
    "\n",
    "    def flip(self):\n",
    "        self.state = np.rot90(self.state, 2)\n",
    "        return self\n",
    "class Field():\n",
    "\n",
    "    WIDTH = 10\n",
    "    HEIGHT = 20\n",
    "    SCORING_ELEMENTS = 6\n",
    "\n",
    "    def __init__(self, state=None):\n",
    "        \"\"\"\n",
    "        Initializes a Tetris Field.\n",
    "        Rows increase downward and columns increase to the right.\n",
    "        \"\"\"\n",
    "        if state is not None:\n",
    "            self.state = np.array(state, dtype=np.uint8, copy=True)\n",
    "        else:\n",
    "            self.state = np.full((Field.HEIGHT, Field.WIDTH), 0, dtype=np.uint8)\n",
    "\n",
    "    def __str__(self):\n",
    "        \"\"\"\n",
    "        Returns a string representation of the field.\n",
    "        \"\"\"\n",
    "        bar = '   |' + ' '.join(map(str, range(Field.WIDTH))) + '|\\n'\n",
    "        mapped_field = np.vectorize(Tetromino.TYPES.__getitem__)(self.state)\n",
    "        field = '\\n'.join(['{:2d} |'.format(i) +\n",
    "            ' '.join(row) + '|' for i, row in enumerate(mapped_field)])\n",
    "        return bar + field + '\\n' + bar\n",
    "\n",
    "    def _test_tetromino_(self, tetromino, r_start, c_start):\n",
    "        \"\"\"\n",
    "        Tests to see if a tetromino can be placed at the specified row and\n",
    "        column. It performs the test with the top left corner of the\n",
    "        tetromino at the specified row and column.\n",
    "        \"\"\"\n",
    "        r_end, c_end = r_start + tetromino.height(), c_start + tetromino.width()\n",
    "        if c_start < 0 or c_end > Field.WIDTH:\n",
    "            return False\n",
    "        if r_start < 0 or r_end > Field.HEIGHT:\n",
    "            return False\n",
    "        test_area = self.state[r_start:r_end, c_start:c_end]\n",
    "        for s, t in zip(test_area.flat, tetromino.flat()):\n",
    "            if s != 0 and t != 0:\n",
    "                return False\n",
    "        return True\n",
    "\n",
    "    def _place_tetromino_(self, tetromino, r_start, c_start):\n",
    "        \"\"\"\n",
    "        Place a tetromino at the specified row and column.\n",
    "        The bottom left corner of the tetromino will be placed at the specified\n",
    "        row and column. This function does not perform checks and will overwrite\n",
    "        filled spaces in the field.\n",
    "        \"\"\"\n",
    "        r_end, c_end = r_start + tetromino.height(), c_start + tetromino.width()\n",
    "        if c_start < 0 or c_end > Field.WIDTH:\n",
    "            return False\n",
    "        if r_start < 0 or r_end > Field.HEIGHT:\n",
    "            return False\n",
    "        for tr, sr in enumerate(range(r_start, r_end)):\n",
    "            for tc, sc, in enumerate(range(c_start, c_end)):\n",
    "                if tetromino[tr][tc] != 0:\n",
    "                    self.state[sr][sc] = tetromino[tr][tc]\n",
    "\n",
    "    def _get_tetromino_drop_row_(self, tetromino, column):\n",
    "        \"\"\"\n",
    "        Given a tetromino and a column, return the row that the tetromino\n",
    "        would end up in if it were dropped in that column.\n",
    "        Assumes the leftmost column of the tetromino will be aligned with the\n",
    "        specified column.\n",
    "        \"\"\"\n",
    "        if column < 0 or column + tetromino.width() > Field.WIDTH:\n",
    "            return -1\n",
    "        last_fit = -1\n",
    "        for row in range(tetromino.height(), Field.HEIGHT):\n",
    "            if self._test_tetromino_(tetromino, row, column):\n",
    "                last_fit = row\n",
    "            else:\n",
    "                return last_fit\n",
    "        return last_fit\n",
    "    \n",
    "    def copy(self):\n",
    "        \"\"\"\n",
    "        Returns a shallow copy of the field.\n",
    "        \"\"\"\n",
    "        return Field(self.state)\n",
    "\n",
    "    def _line_clear_(self):\n",
    "        \"\"\"\n",
    "        Checks and removes all filled lines.\n",
    "        \"\"\"\n",
    "        filled = np.array([row.all() and row.any() for row in self.state])\n",
    "        num_clear_lines = len(self.state[filled])\n",
    "        non_filled = np.array(\n",
    "            [not row.all() and row.any() for row in self.state])\n",
    "        if non_filled.any():\n",
    "            tmp = self.state[non_filled]\n",
    "            self.state.fill(0)\n",
    "            self.state[Field.HEIGHT - tmp.shape[0]:] = tmp\n",
    "        return num_clear_lines\n",
    "\n",
    "    def drop(self, tetromino, column):\n",
    "        \"\"\"\n",
    "        Drops a tetromino in the specified column.\n",
    "        The leftmost column of the tetromino will be aligned with the specified\n",
    "        column.\n",
    "        Returns the row it was dropped in for computations or -1 if a drop was\n",
    "        unable to be computed.\n",
    "        \"\"\"\n",
    "        assert isinstance(tetromino, Tetromino)\n",
    "        row = self._get_tetromino_drop_row_(tetromino, column)\n",
    "        if row == -1:\n",
    "            return row, -1\n",
    "        self._place_tetromino_(tetromino, row, column)\n",
    "        num_lines_cleared = self._line_clear_()\n",
    "        return row, num_lines_cleared\n",
    "\n",
    "    def count_gaps(self):\n",
    "        \"\"\"\n",
    "        Check each column one by one to make sure there are no gpas in the\n",
    "        column.\n",
    "        \"\"\"\n",
    "        rows_with_holes = set()\n",
    "        well_cells = 0\n",
    "        gaps = 0\n",
    "        filled_rows_above_highest_hole = 0\n",
    "        # Cut off all the empty space above all the placed tetrominos\n",
    "        top_indices = np.argmax(self.state.T != 0, axis = 1)\n",
    "        # Count the number of gaps past the first filled space per column\n",
    "        for col, top in zip(self.state.T, top_indices):\n",
    "            for index in range(top, 20):\n",
    "                if col[index] == 0:\n",
    "                    gaps += 1\n",
    "                    rows_with_holes.add(index)\n",
    "        lst = sorted(rows_with_holes)\n",
    "        if len(rows_with_holes):\n",
    "            highest_hole = 20 - lst[0]\n",
    "        else:\n",
    "            highest_hole = 0\n",
    "\n",
    "        for pos, row in enumerate(self.state):\n",
    "            if pos == highest_hole:\n",
    "                break\n",
    "            elif row.all():\n",
    "                filled_rows_above_highest_hole += 1\n",
    "\n",
    "        top_indices -= 1\n",
    "        for col in range(1,9):\n",
    "            cur_row = top_indices[col]\n",
    "            if not self.state[cur_row][col] and self.state[cur_row][col - 1] and self.state[cur_row][col + 1]:\n",
    "                well_cells += 1\n",
    "        return len(rows_with_holes), gaps, highest_hole, filled_rows_above_highest_hole, well_cells\n",
    "\n",
    "    def heights(self):\n",
    "        \"\"\"\n",
    "        Return an array containing the heights of each column.\n",
    "        \"\"\"\n",
    "        heights = Field.HEIGHT - np.argmax(self.state.T != 0, axis=1)\n",
    "        max_height = np.amax(heights)\n",
    "        total_bumpiness = 0\n",
    "        max_bumpiness = 0\n",
    "        for i in range(9):\n",
    "            bumpiness = abs(heights[i]- heights[i+1])\n",
    "            max_bumpiness = max(bumpiness, max_bumpiness)\n",
    "            total_bumpiness += bumpiness\n",
    "        return max_height, max_bumpiness, total_bumpiness\n",
    "\n",
    "    def get_scoring_vector(self):\n",
    "        \"\"\"\n",
    "        Get a vector of values derived from the field used to score a tetromino\n",
    "        placement.\n",
    "        \"\"\"\n",
    "        max_height, max_bumpiness, total_bumpiness = self.heights()\n",
    "        rows_with_holes, holes, highest_hole, filled_rows_above_highest_hole, well_cells = self.count_gaps()\n",
    "        return np.array([\n",
    "            holes,                          # number of holes\n",
    "            total_bumpiness,                # sum of bumpiness\n",
    "            max_height,                     # maximum height\n",
    "            np.count_nonzero(self.state),   # number of blocks\n",
    "            rows_with_holes,                # rows with holes\n",
    "            max_bumpiness,                  # max bumpiness\n",
    "            well_cells,                     # well cells\n",
    "            highest_hole,                   # highest hole\n",
    "            filled_rows_above_highest_hole  # rows above the highest hole\n",
    "        ])\n",
    "\n",
    "    def get_optimal_drop(self, tetromino, weights=None):\n",
    "        \"\"\"\n",
    "        Given a tetromino and a vector of scoring weights, this method\n",
    "        calculates the best placement of the tetromino, scoring each placement\n",
    "        with the weight vector.\n",
    "        \"\"\"\n",
    "        rotations = [\n",
    "            tetromino,\n",
    "            tetromino.copy().rotate_right(),\n",
    "            tetromino.copy().flip(),\n",
    "            tetromino.copy().rotate_left()\n",
    "        ]\n",
    "\n",
    "        best_field = None\n",
    "        best_drop_score = math.inf\n",
    "        for rotation, tetromino_ in enumerate(rotations):\n",
    "            for column in range(Field.WIDTH):\n",
    "                f = self.copy()\n",
    "                row, num_lines_cleared = f.drop(tetromino_, column)\n",
    "                if row == -1:\n",
    "                    continue\n",
    "                scoring_vector = f.get_scoring_vector()\n",
    "                scoring_vector = np.insert(scoring_vector, 0, num_lines_cleared)\n",
    "                if weights is not None:\n",
    "                    score = scoring_vector.dot(weights)\n",
    "                else:\n",
    "                    score = scoring_vector.sum()\n",
    "                if score < best_drop_score:\n",
    "                    best_drop_score = score\n",
    "                    best_field = f\n",
    "        return best_field, num_lines_cleared\n",
    "GENES = 10\n",
    "\n",
    "class Chromosome():\n",
    "\n",
    "    N_SIMULATIONS = 10\n",
    "    MUTATION_CHANCE = 0.01\n",
    "\n",
    "    @staticmethod\n",
    "    def random():\n",
    "        \"\"\"\n",
    "        Returns a random initialized Chromosome.\n",
    "        \"\"\"\n",
    "        return Chromosome(Chromosome.random_genes())\n",
    "\n",
    "    @staticmethod\n",
    "    def random_genes():\n",
    "        \"\"\"\n",
    "        Returns a randomly initialized array for the genetic values inside\n",
    "        a Chromosome. The genetic values are sampled from the open interval from\n",
    "        -1 to 1.\n",
    "        \"\"\"\n",
    "        return (np.random.random_sample(GENES) * 2) - 1\n",
    "\n",
    "    @staticmethod\n",
    "    def set_globals(n_simulations,mutation_chance):\n",
    "        \"\"\"\n",
    "        Sets the static constants that affect how chromosomes will cross and\n",
    "        be evaluated.\n",
    "        \"\"\"\n",
    "        Chromosome.N_SIMULATIONS = n_simulations\n",
    "        Chromosome.MUTATION_CHANCE = mutation_chance\n",
    "\n",
    "    def __init__(self, genes=None):\n",
    "        \"\"\"\n",
    "        Initializes a Chromosome.\n",
    "        \"\"\"\n",
    "        self.genes = genes\n",
    "        self.fitness = None\n",
    "        self.field_score = None\n",
    "        self.length = None\n",
    "        self.clear = None\n",
    "        self.simulations = Chromosome.N_SIMULATIONS\n",
    "        self.mutation_chance = Chromosome.MUTATION_CHANCE\n",
    "\n",
    "    def __str__(self):\n",
    "        \"\"\"\n",
    "        Returns a string representation of the chromosome's genes.\n",
    "        \"\"\"\n",
    "        return 'Weights{}\\n# of blocks dropped{}\\n# of lines cleared{}\\n Sccore{}'.format(self.genes, self.length, self.clear, self.fitness)\n",
    "\n",
    "    def cross(self, other):\n",
    "        \"\"\"\n",
    "        Performs genetic crossing between this chromosome and another\n",
    "        chromosome, returning a new chromosome.\n",
    "        The genetic values stored in a chromosome are all numeric in the open\n",
    "        interval from -1 to 1. When we cross two chromosomes, we will take the\n",
    "        weighted average between the genetic values of the two chromosomes,\n",
    "        weighting it according to the fitness value of each chromosome.\n",
    "        During chromosomal crossing, a mutation has a chance to occur for each\n",
    "        genetic value. If one occurs, that genetic value will be set to a new\n",
    "        random number in the open interval from -1 to 1.\n",
    "        \"\"\"\n",
    "        w_sum = self.get_fitness() + other.get_fitness()\n",
    "        genes = (self.genes * self.get_fitness() / w_sum) + (\n",
    "            other.genes * other.fitness / w_sum)\n",
    "        # Each genetic value has a chance to mutate.\n",
    "        mutated_genes = np.random.random_sample(GENES) < self.mutation_chance\n",
    "        if np.any(mutated_genes):\n",
    "            nonmutated_genes = mutated_genes == 0\n",
    "            mutation = Chromosome.random_genes()\n",
    "            # Compose the original genetic values with the mutated genetic\n",
    "            # values.\n",
    "            genes = (genes * nonmutated_genes) + (mutation * mutated_genes)\n",
    "        return Chromosome(genes)\n",
    "\n",
    "    def get_fitness(self):\n",
    "        \"\"\"\n",
    "        Returns the fitness of the chromosome. If it is cached, it will fetch\n",
    "        the cached fitness value, otherwise it will compute it and return it.\n",
    "        \"\"\"\n",
    "        if self.fitness is not None:\n",
    "            return self.fitness\n",
    "        self.recalculate_fitness()\n",
    "        return self.fitness\n",
    "\n",
    "    def recalculate_fitness(self):\n",
    "        \"\"\"\n",
    "        Calculates the fitness values of the chromosome by running a number\n",
    "        of Tetris simulations using the chromosome and averaging their\n",
    "        performance.\n",
    "        \"\"\"\n",
    "        scores = np.array([\n",
    "            self._get_fitness_() for _ in range(self.simulations)])\n",
    "        self.length, self.clear = np.mean(scores, axis=0)\n",
    "        self.fitness = self.length + (10 * self.clear * self.clear)\n",
    "\n",
    "    def _get_fitness_(self):\n",
    "        \"\"\"\n",
    "        Helper method to perform a single simulation to evaluate the performance\n",
    "        of the chromosome. This will be called multiple times and the overall\n",
    "        performance of the chromosome is the average of all the runs.\n",
    "        \"\"\"\n",
    "        tetrominos = [\n",
    "            Tetromino.ITetromino(),\n",
    "            Tetromino.OTetromino(),\n",
    "            Tetromino.TTetromino(),\n",
    "            Tetromino.STetromino(),\n",
    "            Tetromino.ZTetromino(),\n",
    "            Tetromino.JTetromino(),\n",
    "            Tetromino.LTetromino()\n",
    "        ]\n",
    "        field = Field()\n",
    "        fitness = 0\n",
    "        length = 0\n",
    "        clear = 0 \n",
    "        while True:\n",
    "            tetromino = random.choice(tetrominos)\n",
    "            _field, cleared_lines = field.get_optimal_drop( tetromino, self.genes)\n",
    "            if _field is None :\n",
    "                break\n",
    "            else:\n",
    "                length += 1\n",
    "                clear += cleared_lines\n",
    "                field = _field\n",
    "                # fitness = length + 10 * clear * clear\n",
    "        # fitness = length + (10 * clear * clear)\n",
    "        return length, clear\n",
    "class Generation():\n",
    "\n",
    "    def __init__(self, population, selection, threshold):\n",
    "        \"\"\"\n",
    "        Initializes a Generation of individuals/chromosomes.\n",
    "        \"\"\"\n",
    "        self.population = population\n",
    "        self.threshold = threshold\n",
    "        self.generations = 0\n",
    "        self.selection = selection\n",
    "\n",
    "    def run(self):\n",
    "        \"\"\"\n",
    "        This method will run the genetic algorithm on the population for the\n",
    "        given number of generations.\n",
    "        \"\"\"\n",
    "        last = None\n",
    "        \n",
    "        for i in range(20):\n",
    "            # Sort the population of chromosomes by their fitness\n",
    "            population_by_fitness = sorted(\n",
    "                self.population, key=lambda gene: gene.get_fitness())\n",
    "            fittest_member = population_by_fitness[-1]\n",
    "            # if last is not None and fittest_member.get_fitness() - last.get_fitness() <= self.threshold:\n",
    "            #     return last\n",
    "            # else:\n",
    "            #     last = fittest_member\n",
    "            print('Generation: {}'.format(self.generations))\n",
    "            print([member.get_fitness() for member in population_by_fitness])\n",
    "            print('Fittest member: {}'.format(fittest_member))\n",
    "            print('')\n",
    "            # calculate how many members should be removed\n",
    "            length = len(population_by_fitness)\n",
    "            cut = math.floor(length * self.selection) \n",
    "            # Memebers in the population after selection\n",
    "            fittest = population_by_fitness[cut:]\n",
    "            # Shuffle and cross breed the fittest members.\n",
    "            random.shuffle(fittest)\n",
    "            # new_gen = []\n",
    "            for i in range(0, length - cut, 2):\n",
    "                if i+1 < length - cut:\n",
    "                    fittest.append(fittest[i].cross(fittest[i + 1]))\n",
    "                # else:\n",
    "                #     new_gen.append(fittest[i])\n",
    "            self.population = fittest\n",
    "            for chromosome in self.population:\n",
    "                chromosome.recalculate_fitness()\n",
    "            self.generations += 1\n",
    "        return sorted(self.population, key=lambda gene: gene.get_fitness())[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Chromosome.set_globals(10,0.1) # initialize macro for N and q\n",
    "population = Generation([Chromosome.random() for i in range(25)], 0.4, 0) \n",
    "# initialize macro for p and stopping_criteria\n",
    "fittest = population.run()\n",
    "print('Fittest member: {}'.format(fittest))"
   ]
  }
 ]
}